File: lectures/lec6/main.cc
#include<iostream>
#include<cmath>
#include<complex>
#include<string>
#include<functional>

typedef std::complex<double> complex;

template < typename T >
void echo(T arg){
    std::cout << "echo: " << arg << "\n";
}

void fun_of_two(std::function<double(double)> f){std::cout << "f(2) = " << f(2) << "\n";}

int main(){
    echo("string");
    echo(1);
    echo(1.23);
    complex I = std::sqrt((complex)(-1));
    echo(std::sin(std::sqrt(I)));

    std::string s="original string";
    std::function<void(void)> prints = [=, &s](){std::cout << "prints: s = " << s << "\n";};  //[=, &s] betyder at alle varialble der ikke er s gemmes som værdi, imens s gemmes som reference 
    // std::function<void(void)> prints = [=](){std::cout << "prints: s = " << s << "\n";};
    // std::function<void(void)> prints = [&](){std::cout << "prints: s = " << s << "\n";};
    prints();
    s="new string";
    prints();

    std::function<double(double)> f = [](double x){return x*x;};
    fun_of_two(f);
    fun_of_two([](double x){return x*x;});

    return 0;
}

File: lectures/lec6/main.cs
using static System.Console;
using complex=System.Numerics.Complex;
class main{
    static void echo<T>(T arg){
        WriteLine($"echo: {arg}");
    }
    // static void addition<T>(T x, T y){
    //     WriteLine($"echo: {x+y}");
    // }
    static void fun_of_one(System.Func<double, double> f){ WriteLine($"f(2)={f(2)}");}
    static int Main(){
        System.Func<double, double> f = delegate(double x){return x*x;};
        f = x => x*x;
        f = System.Math.Sin;
        WriteLine($"f: {f(3)}");

        fun_of_one(f);
        fun_of_one(x => x*x);


        echo("string");
        echo(1);
        echo(1.3);
        complex I = complex.Sqrt(-1);
        echo(complex.Sqrt(-1));
        echo(complex.Pow(I,I));
        echo(complex.Sin(I));

        string s="original string";
        System.Action prints = delegate(){WriteLine($"prints: s={s}");};
        prints();
        s="new string";
        prints();

        return 0;
    }
}